src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb

```diff
@@ -3429,61 +3429,174 @@ class C
+                                                                         Finally
+                                                                             ' 4. Unblock the first task and the main thread.
+                                                                             tcs.SetResult(True)
+                                                                         End Try
+                                                                     End Sub)
+                                                End Sub
+
+                    Try
+                        AddHandler provider.ProviderCalled, providerCalledHandler
+
+                        ' In the new completion, when pressed <ctrl>-<space>, we have to wait for the aggregate operation to complete.
+                        ' 1. Hang here.
+                        state.SendCommitUniqueCompletionListItem()
+
+                        Assert.NotNull(task1)
+                        Assert.NotNull(task2)
+                        Await Task.WhenAll(task1, task2)
+
+                        Await state.WaitForAsynchronousOperationsAsync()
+                        Await state.AssertNoCompletionSession()
+                        Assert.Contains("System", state.GetLineTextFromCaretPosition())
+                    Finally
+                        RemoveHandler provider.ProviderCalled, providerCalledHandler
+                    End Try
+                End If
             End Using
+
         End Function
 
-        <InlineData(CompletionImplementation.Legacy)>
-        <InlineData(CompletionImplementation.Modern, Skip:="https://github.com/dotnet/roslyn/issues/29112")>
+        <MemberData(NameOf(AllCompletionImplementations))>
         <WpfTheory, Trait(Traits.Feature, Traits.Features.Completion)>
         Public Async Function TestNoBlockOnCompletionItems3(completionImplementation As CompletionImplementation) As Task
+            ' This test verifies a scenario with the following conditions:
+            ' a. A slow completion provider
+            ' b. The block option set to false.
+            ' Scenario:
+            ' 1. Type 'Sys'
+            ' 2. Send CommitIfUnique (Ctrl + space)
+            ' 3. Wait for 250ms.
+            ' 4. Verify that there is no completion window shown. In the new completion, we can just start the verification and check that the verification is still running.
+            ' 5. Check that the commit is not yet provided: there is 'Sys' but no 'System'
+            ' 6. The next statement in the UI thread after CommitIfUnique is typing 'a'.
+            ' 7. Simulate unblocking the provider.
+            ' 8. Verify that 
+            ' 8.a. The old completion adds 'a' to 'Sys' and displays 'Sysa'. CommitIfUnique is canceled because it was interrupted by typing 'a'.
+            ' 8.b. The new completion completes CommitIfUnique and then add 'a'.
```

Comment:
I can't remember all the details of our discussion.  We're going with 8.b. because 8.a. is impossible in the new system?

Reply:
Yes, in the new system, if we would cancel it on on each immediate keypress, it seems to be inconsistent.
Therefore, if we want to be able to cancel it after some time, it is difficult to define a valid threshold: do not cancel if pressed before, and cancel if pressed after it.